src/raqm.c

```diff
@@ -377,14 +385,29 @@ raqm_destroy (raqm_t *rq)
   if (!rq || --rq->ref_count != 0)
     return;
 
-  free (rq->text);
-  free (rq->text_utf8);
-  _raqm_free_text_info (rq);
-  _raqm_free_runs (rq);
-  free (rq->glyphs);
+  _raqm_free_intermediate_data (rq);
   free (rq);
 }
 
+/**
+ * raqm_clear_contents:
+ * @rq: a #raqm_t.
+ *
+ * Clears internal state of previously used raqm_t object, making it
+ * ready for reuse.
+ *
+ * Since: 0.9
+ */
+void
+raqm_clear_contents (raqm_t *rq)
+{
+  if (!rq)
+    return;
+
+  _raqm_free_intermediate_data (rq);
```

Comment:
Ideally, we shouldn't be freeing memory here but allow allocated memory to be reused (and reallocating if we need more than what is currently allocated), otherwise reusing `raqm_t` is not much different from creating new one every time.

Reply:
Good point. I wanted to do that but seeing a [temporary buffer](https://github.com/HOST-Oman/libraqm/blob/ced707f86c454baa30b3c2b056cc69bd4a24d3db/src/raqm.c#L518) for UTF-8 -> UTF-32 I have got an impression that you target the least memory consumption possible and we should shrink to fit every time.

Even in this case reusing has benefits:
- At least one alloc/free is saved per use. Not much but still good.
- Uniform settings are kept so you haven't to setup the new raqm_t from scratch.

Still I will be happy to implement memory reuse, just take me some time) And do you have an idea at which threshold should we shrink? E.g. if the new text is at least twice shorter than the buffer.

Reply:
Thinking further I came up with an idea that there should not be auto-shrinking at all. E.g. if you process the text paragraph by paragraph and have two very long ones splitted by the very short one, shrinking will be more of harm than profit. And this is not rare case.

So let us always keep allocated memory between raqm_t reuses, and force usage of raqm_destroy if a user wants to free memory. That will be his/her responsibility to decide when this is good.

Reply:
Yes, that would be my preference as well.

Reply:
Done